lib/rex/proto/dns/custom_nameserver_provider.rb

```diff
@@ -0,0 +1,234 @@
+module Rex
+module Proto
+module DNS
+
+  ##
+  # Provides a DNS resolver the ability to use different nameservers
+  # for different requests, based on the domain being queried.
+  ##
+  module CustomNameserverProvider
+    CONFIG_KEY = 'framework/dns'
+
+    #
+    # A Comm implementation that always reports as dead, so should never
+    # be used. This is used to prevent DNS leaks of saved DNS rules that 
+    # were attached to a specific channel.
+    ##
+    class CommSink
+      include Msf::Session::Comm
+      def alive?
+        false
+      end
+
+      def supports_udp?
+        # It won't be used anyway, so let's just say we support it
+        true
+      end
+
+      def sid
+        'previous MSF session'

```

Comment:
Just curious - why was this needed?

Reply:
Having the sink-holed DNS class was to satisfy the condition described above:

> If a specific session was specified for a particular DNS rule, and then that session dies, DNS resolution will not fall back to custom or default nameservers, since we interpret the user's intentional use of a particular session to mean "I don't want DNS to fallback to other nameservers and potentially leak". Falling back upon a session closure would violate this.
> 
> This is also applicable when the application closes and re-opens. If the user's DNS configuration was saved, and that configuration had included a specific session to route certain DNS requests through, MSF will remember that there was a session associated with the rule, and that rule will be treated as a black hole: these DNS requests will deliberately fail, and not fall back.